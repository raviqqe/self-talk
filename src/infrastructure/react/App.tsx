import { PulseLoader } from "react-spinners";
import { useAsync } from "react-use";
import styled from "styled-components";
import { Home, IProps as IHomeProps } from "./Home";
import { IProps as ILandingProps, Landing } from "./Landing";
import { white } from "./style/colors";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
`;

export interface IProps extends IHomeProps, ILandingProps {
  initialize: () => Promise<void>;
  signedIn: boolean | null;
}

export const App = ({
  documents,
  initialize,
  repositoryURL,
  signedIn,
  signIn,
  signOut,
  ...props
}: IProps): JSX.Element => {
  useAsync(initialize, []);

  return signedIn === null ? (
    <LoaderContainer>
      <PulseLoader color={white} />
    </LoaderContainer>
  ) : signedIn ? (
    <Home {...props} documents={documents} signOut={signOut} />
  ) : (
    <Landing repositoryURL={repositoryURL} signIn={signIn} />
  );
};
