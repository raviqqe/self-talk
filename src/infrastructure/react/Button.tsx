import { defaultImport } from "default-import";
import defaultStyled from "styled-components";
import { lightGrey, red, white } from "./style/colors.js";
import { boxShadow } from "./style/index.js";

const styled = defaultImport(defaultStyled);

export const Button = styled.button<{ secondary?: boolean }>`
  ${boxShadow};
  background: ${({ secondary }) => (secondary ? lightGrey : red)};
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: inherit;
  color: ${white};
  border: none;
  cursor: pointer;
  flex-shrink: 0;
`;
