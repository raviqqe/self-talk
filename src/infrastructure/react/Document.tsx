import { MdEdit } from "react-icons/md";
import React, { useState } from "react";
import styled from "styled-components";
import { IconButton } from "./IconButton";
import { Markdown } from "./Markdown";
import {
  UpdateDocument,
  IProps as IUpdateDocumentProps,
} from "./UpdateDocument";
import { boxShadow } from "./style";
import { white } from "./style/colors";

const Container = styled.div`
  ${boxShadow};
  background: ${white};
  padding: 1em;
  padding-right: 2.1em;
  border-radius: 0.5em;
  position: relative;
`;

const ButtonContainer = styled.div`
  display: flex;
  position: absolute;
  top: 0.4em;
  right: 0.4em;
`;

interface IProps extends IUpdateDocumentProps {}

export const Document = ({
  document,
  insertFiles,
  updateDocument,
  ...restProps
}: IProps) => {
  const [editing, setEditing] = useState(false);

  return editing ? (
    <UpdateDocument
      document={document}
      insertFiles={insertFiles}
      updateDocument={async (document) => {
        setEditing(false);
        await updateDocument(document);
      }}
      {...restProps}
    />
  ) : (
    <Container {...restProps}>
      <Markdown>{document.text}</Markdown>
      <ButtonContainer>
        <IconButton aria-label="Edit" onClick={() => setEditing(true)}>
          <MdEdit />
        </IconButton>
      </ButtonContainer>
    </Container>
  );
};
