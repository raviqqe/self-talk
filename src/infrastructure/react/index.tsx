import { createRoot, Root } from "react-dom/client";
import { ApplicationInitializer } from "../../application/application-initializer";
import { DocumentCreator } from "../../application/document-creator";
import { DocumentLister } from "../../application/document-lister";
import { DocumentUpdater } from "../../application/document-updater";
import { SignInManager } from "../../application/sign-in-manager";
import { SignOutManager } from "../../application/sign-out-manager";
import { TextFileInserter } from "../../application/text-file-inserter";
import { IDocument } from "../../domain/document";
import { IRenderer } from "../renderer";
import { App, IProps as IAppProps } from "./App";
import { GlobalStyle } from "./style";

interface IPresenter {
  setRenderer(renderer: IRenderer): void;
}

interface IProps extends Pick<IAppProps, "documents" | "signedIn"> {}

export class ReactRenderer implements IRenderer {
  private readonly root: Root;
  private props: IProps = { documents: null, signedIn: null };

  constructor(
    element: HTMLElement,
    presenters: IPresenter[],
    private readonly applicationInitializer: ApplicationInitializer,
    private readonly documentCreator: DocumentCreator,
    private readonly documentLister: DocumentLister,
    private readonly documentUpdater: DocumentUpdater,
    private readonly signInManager: SignInManager,
    private readonly signOutManager: SignOutManager,
    private readonly textFileInserter: TextFileInserter,
    private readonly repositoryURL: string
  ) {
    for (const presenter of presenters) {
      presenter.setRenderer(this);
    }

    this.root = createRoot(element);
  }

  public render(): void {
    this.renderProps({});
  }

  public renderDocuments(documents: IDocument[] | null): void {
    this.renderProps({ documents });
  }

  public renderSignedIn(signedIn: boolean): void {
    this.renderProps({ signedIn });
  }

  private renderProps(props: Partial<IProps>): void {
    this.props = { ...this.props, ...props };

    this.root.render(
      <>
        <App
          {...this.props}
          createDocument={(text: string) => this.documentCreator.create(text)}
          initialize={() => this.applicationInitializer.initialize()}
          insertFiles={(
            text: string,
            position: number,
            files: File[]
          ): Promise<string> =>
            this.textFileInserter.insert(text, position, files)
          }
          listDocuments={() => this.documentLister.list()}
          listMoreDocuments={() => this.documentLister.listMore()}
          repositoryURL={this.repositoryURL}
          signIn={() => this.signInManager.signIn()}
          signOut={() => this.signOutManager.signOut()}
          updateDocument={(document: IDocument) =>
            this.documentUpdater.update(document)
          }
        />
        <GlobalStyle />
      </>
    );
  }
}
