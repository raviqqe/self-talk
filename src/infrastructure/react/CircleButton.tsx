import styled from "styled-components";
import { Button } from "./Button.js";

export const CircleButton = styled(Button)`
  font-size: 1.5em;
  padding: 0;
  height: 2em;
  width: 2em;
  border-radius: 1em;
`;
