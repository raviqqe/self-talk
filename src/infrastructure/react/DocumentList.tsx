import { defaultImport } from "default-import";
import { PulseLoader } from "react-spinners";
import { useAsync } from "react-use";
import defaultStyled from "styled-components";
import type * as domain from "../../domain.js";
import { Document } from "./Document.js";
import { white } from "./style/colors.js";
import { type InsertFilesFunction } from "./utilities.js";
import defaultUseInfiniteScroll from "react-infinite-scroll-hook";

const useInfiniteScroll = defaultImport(defaultUseInfiniteScroll);
const styled = defaultImport(defaultStyled);

const Container = styled.div`
  display: flex;
  flex-direction: column-reverse;
  overflow: auto;
  padding: 1em 0.5em;
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 1rem;
`;

const StyledDocument = styled(Document)`
  margin: 0.5em;
`;

export interface Props {
  documents: domain.Document[] | null;
  insertFiles: InsertFilesFunction;
  listDocuments: () => Promise<void>;
  listMoreDocuments: () => Promise<void>;
  updateDocument: (document: domain.Document) => Promise<void>;
}

export const DocumentList = ({
  documents,
  insertFiles,
  listDocuments,
  listMoreDocuments,
  updateDocument,
}: Props): JSX.Element => {
  useAsync(listDocuments, []);

  const [ref] = useInfiniteScroll({
    loading: false,
    hasNextPage: true,
    onLoadMore: listMoreDocuments,
  });

  return (
    <Container>
      {documents?.map((document) => (
        <StyledDocument
          document={document}
          insertFiles={insertFiles}
          key={document.id}
          updateDocument={updateDocument}
        />
      ))}
      <LoaderContainer ref={ref}>
        <PulseLoader color={white} />
      </LoaderContainer>
    </Container>
  );
};
