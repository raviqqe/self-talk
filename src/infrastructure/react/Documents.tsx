import { PulseLoader } from "react-spinners";
import InfiniteScroller from "react-infinite-scroller";
import { useAsync } from "react-use";
import React from "react";
import styled from "styled-components";
import { IDocument } from "../../domain/document";
import { InsertFilesFunction } from "./utilities";
import { Document } from "./Document";
import { white } from "./style/colors";

const Container = styled.div`
  display: flex;
  flex-direction: column-reverse;
  overflow: auto;
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const StyledInfiniteScroller = styled(InfiniteScroller)`
  display: flex;
  flex-direction: column-reverse;
  padding: 1em 0.5em;
`;

const StyledDocument = styled(Document)`
  margin: 0.5em;
`;

export interface IProps {
  documents: IDocument[] | null;
  insertFiles: InsertFilesFunction;
  listDocuments: () => Promise<void>;
  listMoreDocuments: () => Promise<void>;
  updateDocument: (document: IDocument) => Promise<void>;
}

export const Documents = ({
  documents,
  insertFiles,
  listDocuments,
  listMoreDocuments,
  updateDocument,
}: IProps) => {
  useAsync(listDocuments, []);

  return documents ? (
    <Container>
      <StyledInfiniteScroller
        hasMore={true}
        isReverse={true}
        loadMore={listMoreDocuments}
        threshold={512}
        useWindow={false}
      >
        {documents.map((document: IDocument) => (
          <StyledDocument
            document={document}
            insertFiles={insertFiles}
            key={document.id}
            updateDocument={updateDocument}
          />
        ))}
      </StyledInfiniteScroller>
    </Container>
  ) : (
    <LoaderContainer>
      <PulseLoader color={white} />
    </LoaderContainer>
  );
};
