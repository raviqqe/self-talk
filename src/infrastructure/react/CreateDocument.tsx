import { useState } from "react";
import { MdAdd } from "react-icons/md";
import styled from "styled-components";
import { CircleButton } from "./CircleButton";
import { MarkdownTextArea } from "./MarkdownTextArea";
import { InsertFilesFunction } from "./utilities";

const Container = styled.div`
  display: flex;
  align-items: center;
  padding: 1em;
`;

const StyledMarkdownTextArea = styled(MarkdownTextArea)`
  margin-right: -1em;
  max-height: 80vh;
`;

interface IProps {
  createDocument: (text: string) => Promise<void>;
  insertFiles: InsertFilesFunction;
}

export const CreateDocument = ({
  createDocument,
  insertFiles,
}: IProps): JSX.Element => {
  const [text, setText] = useState("");
  const onSubmit = async (): Promise<void> => {
    setText("");
    await createDocument(text);
  };

  return (
    <Container>
      <StyledMarkdownTextArea
        insertFiles={insertFiles}
        onSubmit={onSubmit}
        setText={setText}
        text={text}
      />
      <CircleButton aria-label="Create" onClick={onSubmit}>
        <MdAdd />
      </CircleButton>
    </Container>
  );
};
