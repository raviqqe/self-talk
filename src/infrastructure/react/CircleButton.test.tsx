import { render } from "@testing-library/react";
import { expect, it } from "vitest";
import { CircleButton } from "./CircleButton";

it("renders", () => {
  expect(
    render(<CircleButton>foo</CircleButton>).container.firstChild
  ).toMatchSnapshot();
});

it("renders as a secondary button", () => {
  expect(
    render(<CircleButton secondary={true}>foo</CircleButton>).container
      .firstChild
  ).toMatchSnapshot();
});
