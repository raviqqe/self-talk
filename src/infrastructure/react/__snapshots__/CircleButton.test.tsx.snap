// Vitest Snapshot v1

exports[`renders 1`] = `
<button
  class="sc-bcXHqe sc-gswNZR eTvdtE gWKlTj"
>
  foo
</button>
`;

exports[`renders as a secondary button 1`] = `
<button
  class="sc-bcXHqe sc-gswNZR fmEPPj gWKlTj"
>
  foo
</button>
`;
